src/app/qgsgeometryvalidationdock.cpp

```diff
@@ -0,0 +1,343 @@
+  if ( !featureExtent.isEmpty() )
+  {
+    QgsRectangle mapExtent = mLayerTransform.transform( featureExtent );
+    mMapCanvas->zoomToFeatureExtent( mapExtent );
+  }
+}
+
+void QgsGeometryValidationDock::triggerTopologyChecks()
+{
+  QgsVectorLayer *layer = qobject_cast<QgsVectorLayer *>( mMapCanvas->currentLayer() );
+  if ( layer )
+    mGeometryValidationService->triggerTopologyChecks( layer );
+}
+
+void QgsGeometryValidationDock::updateLayerTransform()
+{
+  if ( !mMapCanvas->currentLayer() )
+    return;
+
+  mLayerTransform = QgsCoordinateTransform( mMapCanvas->currentLayer()->crs(), mMapCanvas->mapSettings().destinationCrs(), mMapCanvas->mapSettings().transformContext() );
+}
+
+void QgsGeometryValidationDock::onDataChanged( const QModelIndex &topLeft, const QModelIndex &bottomRight, const QVector<int> &roles )
+{
+  Q_UNUSED( bottomRight )
+  Q_UNUSED( roles )
+
+  if ( currentIndex() == topLeft )
+    updateCurrentError();
+}
+
+void QgsGeometryValidationDock::onRowsInserted()
+{
+  if ( !isVisible() )
+  {
+    mQgisApp->addDockWidget( Qt::RightDockWidgetArea, this );
+  }
+  raise();
+}
+
+QgsGeometryValidationService *QgsGeometryValidationDock::geometryValidationService() const
+{
+  return mGeometryValidationService;
+}
+
+void QgsGeometryValidationDock::setGeometryValidationService( QgsGeometryValidationService *geometryValidationService )
+{
+  mGeometryValidationService = geometryValidationService;
+}
+
+void QgsGeometryValidationDock::updateCurrentError()
+{
+  if ( mGeometryValidationModel->rowCount() == 0 )
+    mErrorListView->selectionModel()->clearCurrentIndex();
+
+  mFeatureRubberband->hide();
+  mFeatureRubberband->update();
+  mErrorRubberband->hide();
+  mErrorRubberband->update();
+  mErrorLocationRubberband->hide();
+  mErrorLocationRubberband->update();
+
+  onCurrentErrorChanged( currentIndex(), QModelIndex() );
+}
+
+QModelIndex QgsGeometryValidationDock::currentIndex() const
+{
+  if ( !mGeometryValidationModel->rowCount() )
+    return QModelIndex();
+  else
+    return mErrorListView->selectionModel()->currentIndex();
+}
+
+void QgsGeometryValidationDock::onCurrentErrorChanged( const QModelIndex &current, const QModelIndex &previous )
+{
+  Q_UNUSED( previous )
+
+  mProblemDetailWidget->setVisible( current.isValid() );
+
+  mNextButton->setEnabled( current.isValid() && current.row() < mGeometryValidationModel->rowCount() - 1 );
+  mPreviousButton->setEnabled( current.isValid() && current.row() > 0 );
+
+  mProblemDetailWidget->setVisible( current.isValid() );
+
+  if ( !current.isValid() )
+    return;
+
+  mProblemDescriptionLabel->setText( current.data( QgsGeometryValidationModel::DetailsRole ).toString() );
+
+  QgsGeometryCheckError *error = current.data( QgsGeometryValidationModel::GeometryCheckErrorRole ).value<QgsGeometryCheckError *>();
+  if ( error )
+  {
+    while ( QWidget *w = mResolutionWidget->findChild<QWidget *>() )
+      delete w;
```

Comment:
w->deleteLater() may be safer?

Reply:
not sure, deleting them and start from scratch with the new widgets sounds quite safe to me, so unless there are issues, I'd rather keep it this way.